analysis/plot.routes.all_region_pairs.py

```diff
@@ -0,0 +1,196 @@
+    """
+    Extract the source and destination regions from the given filename.
+    Assumes the format: 'routes.[src].[dst].by_iso.distribution'
+    """
+    regex = re.compile(r'^routes\.([a-z0-9-]+)\.([a-z0-9-]+)\.([a-z0-9-]+)\.([a-z0-9-]+)\.by_(?:\w+)\.distribution$')
+    m = regex.match(filename)
+    assert m is not None, 'Invalid filename: {}'.format(filename)
+    src_cloud = m.group(1)
+    src_region = m.group(2)
+    dst_cloud = m.group(3)
+    dst_region = m.group(4)
+    src = f'{src_cloud}:{src_region}'
+    dst = f'{dst_cloud}:{dst_region}'
+    return src, dst
+
+def plot_pdf(values, weights, src_region: str, dst_region: str, metric: str, data_source: str):
+    # Normalize the weights to ensure they sum up to 1
+    weights = np.array(weights) / np.sum(weights)
+
+    plt.hist(values, bins=len(values), weights=weights, density=True, edgecolor='black')
+    plt.xlabel('Values')
+    plt.ylabel('Probability')
+    plt.title(f'PDF of average {metric} between regions {src_region} and {dst_region}')
+
+    filename = f'{metric}.pdf.{data_source}.{src_region}.{dst_region}.png'
+    logging.info(f'Saving heatmap to {filename} ...')
+    plt.savefig(filename, bbox_inches='tight')
+
+def plot_heatmap(src_regions: list[str], dst_regions: list[str], region_hop_counts: dict[tuple[str, str], float],
+                 metric: str, data_source: str):
+    # Create a mapping of regions to indices
+    src_region_to_index = {region: index for index, region in enumerate(src_regions)}
+    dst_region_to_index = {region: index for index, region in enumerate(dst_regions)}
+
+    # Initialize a matrix for the data
+    matrix = np.zeros((len(src_regions), len(dst_regions)))
+
+    # Populate the matrix
+    for (src_region, dst_region), value in region_hop_counts.items():
+        i, j = src_region_to_index[src_region], dst_region_to_index[dst_region]
+        matrix[i][j] = value
+
+    # Plotting the heatmap
+    plt.figure(figsize=(10, 8))
+    plt.imshow(matrix, cmap='coolwarm', interpolation='nearest')
+    plt.colorbar()
+
+    # Setting labels
+    plt.xticks(ticks=np.arange(len(dst_regions)), labels=dst_regions, rotation=90)
+    plt.yticks(ticks=np.arange(len(src_regions)), labels=src_regions)
+
+    plt.title(f'Average {metric} between regions {data_source}')
+    filename = f'{metric}.heatmap.{data_source}.png'
+    logging.info(f'Saving heatmap to {filename} ...')
+    plt.savefig(filename, bbox_inches='tight')
+
+def get_weighted_average_by_region_pair(dirpath: str, process_hops: Callable[[list[str]], Any],
+                                        metric: str, plot_pdfs: bool):
+    # Dictionary to hold the source-destination pairs and the aggregated value
+    weighted_average_by_region_pair = {}
+
+    # Corrected processing of each file
+    for file in os.listdir(dirpath):
+        file_path = os.path.join(dirpath, file)
+        src, dst = extract_regions_from_filename(file)
+
+        # optional filtering on regions
```

Comment:
Probably just delete this.

Reply:
I guess I'll make this an argument and update the instrucitons. This is actually used in certain cases.

Reply:
Fixed.